engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/IOMappingHelper.java

```diff
@@ -45,34 +48,56 @@
       variablesState.removeTemporaryVariables(elementInstanceKey);
     }
 
-    if (hasOutputMappings) {
-      mergeTool.reset();
-
-      final DirectBuffer variables =
-          determineVariables(variablesState, elementInstanceKey, outputMappings);
-
-      mergeTool.mergeDocumentStrictly(variables, outputMappings);
-      final DirectBuffer mergedVariables = mergeTool.writeResultToBuffer();
+    if (!outputMappings.isEmpty()) {
+      final DirectBuffer mergedVariables =
+          mergeOutputVariables(variablesState, elementInstanceKey, outputMappings);
 
       variablesState.setVariablesFromDocument(
           getVariableScopeKey(context), workflowKey, mergedVariables);
     }
   }
 
+  private DirectBuffer mergeOutputVariables(
+      final VariablesState variablesState,
+      final long elementInstanceKey,
+      final Mappings outputMappings) {
+
+    mergeTool.reset();
+
+    // fetch the variables from the target mappings to allow merging with existing variables
+    final DirectBuffer targetVariables =
+        variablesState.getVariablesAsDocument(
+            elementInstanceKey, outputMappings.getTargetVariableNames());
+
+    // copy variables to not override them with next variable state access
+    buffer.putBytes(0, targetVariables, 0, targetVariables.capacity());
+    bufferView.wrap(buffer, 0, targetVariables.capacity());
+
+    mergeTool.mergeDocument(bufferView);
```

Comment:
this is the old index right ? :sweat_smile:

Reply:
What do you mean with old index? 
By calling `mergeDocument(...)` without any mapping, it passes all variables to the result document. Then merging the other variables on top.

Reply:
In the past the method was called index. For me it looks weird  that it is called merge but anyhow.

Reply:
Agree. A good topic for another day :sweat_smile: